client/src/main/java/io/confluent/kafka/schemaregistry/AbstractSchemaProvider.java

```diff
@@ -48,13 +48,13 @@ protected Map<String, String> resolveReferences(Schema schema) {
       return Collections.emptyMap();
     }
     Map<String, String> result = new LinkedHashMap<>();
-    Set<String> visited = new HashSet<>();
-    resolveReferences(schema, result, visited);
+    resolveReferences(schema, result);
     return result;
   }
 
-  private void resolveReferences(Schema schema, Map<String, String> schemas, Set<String> visited) {
+  private void resolveReferences(Schema schema, Map<String, String> schemas) {
     List<SchemaReference> references = schema.getReferences();
+    Set<String> visited = new HashSet<>();
```

Comment:
I think this may not work with cyclic (recursive) references

Reply:
Ah. Good catch.

Reply:
Would only moving this block outside of the visited check work to fix this problem?
```
      if (reference.getVersion() == -1) {
        // Update the version with the latest
        reference.setVersion(s.getVersion());
      }
```

Reply:
@sriman-confluent Yes. That should help.

@rayokota How can we registry a circular reference. Wouldn't that fail given we cannot find the schema?

Register Test1 with reference Test2 -> Test3 -> Test1. But when Test1 is resolved it will fail as it is not registerd yet?

Reply:
There may already be tests for circular references, so let's see if you can get a green build first.  I think you need to rebase this PR to get the tests to pass.

Reply:
I was able to add a test to validate the scenario